launch_ros/launch_ros/parameter_descriptions.py

```diff
@@ -154,3 +163,96 @@ def evaluate(self, context: LaunchContext) -> Tuple[Text, 'EvaluatedParameterVal
+        """
+        ensure_argument_type(
+            param_file,
+            SomeSubstitutionsType_types_tuple + (os.PathLike, bytes),
+            'param_file',
+            'ParameterFile()'
+        )
+        ensure_argument_type(
+            allow_substs,
+            bool,
+            'allow_subst',
+            'ParameterFile()'
+        )
+        self.__param_file: Union[List[Substitution], FilePath] = param_file
+        if isinstance(param_file, SomeSubstitutionsType_types_tuple):
+            self.__param_file = normalize_to_list_of_substitutions(param_file)
+        self.__allow_substs = allow_substs
+        self.__evaluated_param_file: Optional[Path] = None
+        self.__created_tmp_file = False
+
+    @property
+    def param_file(self) -> Union[FilePath, List[Substitution]]:
+        """Getter for parameter file."""
+        if self.__evaluated_param_file is not None:
+            return self.__evaluated_param_file
+        return self.__param_file
+
+    @property
+    def allow_substs(self) -> bool:
+        """Getter for allow substitutions argument."""
+        return self.__allow_substs
+
+    def __str__(self) -> Text:
+        return (
+            'launch_ros.description.ParameterFile'
+            f'(param_file={self.param_file}, allow_substs={self.allow_substs})'
+        )
+
+    def evaluate(self, context: LaunchContext) -> Path:
+        """Evaluate and return a parameter file path."""
+        if self.__evaluated_param_file is not None:
+            return self.__evaluated_param_file
+
+        param_file = self.__param_file
+        if isinstance(param_file, list):
+            # list of substitutions
+            param_file = perform_substitutions(context, self.__param_file)
+        param_file_path: Path = Path(param_file)
+        if self.__allow_substs:
+            with ExitStack() as stack:
+                f = stack.enter_context(open(param_file_path, 'r'))
+                h = stack.enter_context(
+                    NamedTemporaryFile(mode='w', prefix='launch_params_', delete=False))
+                read = f.read()
+                parsed = perform_substitutions(context, parse_substitution(read))
+                yaml_file_is_valid = True
+                try:
+                    yaml.safe_load(parsed)
+                except Exception:
+                    yaml_file_is_valid = False
+                if not yaml_file_is_valid:
+                    raise SubstitutionFailure(
+                        'The substituted parameter file is not a valid yaml file')
+                h.write(parsed)
+                param_file_path = Path(h.name)
+                self.__created_tmp_file = True
+        self.__evaluated_param_file = param_file_path
+        return param_file_path
+
+    def cleanup(self):
+        """Delete created temporary files."""
+        if self.__evaluated_param_file is not None and self.__created_tmp_file:
+            os.unlink(self.__evaluated_param_file)
```

Comment:
@ivanpauno consider checking that the file still exists OR handling the exception that would result if it doesn't.

Reply:
Also, while we don't handle this correctly elsewhere either, it occurred to me that we may want to hook up file removal to node shutdown or launch shutdown instead of relying on class destructors.

Reply:
> Also, while we don't handle this correctly elsewhere either, it occurred to me that we may want to hook up file removal to node shutdown or launch shutdown instead of relying on class destructors.

It's a bit hard to handle, considering how we currently do normalization/evaluation.
We were already creating temporal files that are [never deleted](https://github.com/ros2/launch_ros/blob/9250313932c2c7477a982ef0c0cfe5124d3cfddf/launch_ros/launch_ros/actions/node.py#L325).

Reply:
> @ivanpauno consider checking that the file still exists OR handling the exception that would result if it doesn't.

Done in https://github.com/ros2/launch_ros/pull/168/commits/a070f45fa96a7178ce87fae316682beb784b8507

Reply:
> We were already creating temporal files that are never deleted.

Yeah, I know. We should do better if we can though.

About https://github.com/ros2/launch_ros/commit/a070f45fa96a7178ce87fae316682beb784b8507, I know I suggested checking for file existence but now I realize I was wrong and that still exposes you to TOCTTOU races. To unlink and deal with the fall out would be the right thing to do.

Reply:
Good point, fixed in https://github.com/ros2/launch_ros/pull/168/commits/eee9e4c8212dd443a200a2ec8ae0f465585aa957.